File: macos/Sources/FloatingLyricCore/Auth/LoginPrompt.swift
import Foundation

/// Decides whether the login window is owed to the user, and which of its
/// three faces to show.
///
/// They are genuinely different moments. Someone opening the app for the first
/// time should meet a single button, not a walkthrough for building a Spotify
/// app — that walkthrough only exists for the rare person supplying their own
/// Client ID.
public enum LoginPrompt: Equatable {
    /// No Client ID at all: show the full walkthrough.
    case setup
    /// First run with a Client ID available: welcome them in.
    case welcome
    /// Signed in before, signed out now: just offer to log back in.
    case logIn

    public static func required(clientID: String?,
                                isLoggedIn: Bool,
                                hasSignedInBefore: Bool) -> LoginPrompt? {
        guard let clientID, !clientID.trimmingCharacters(in: .whitespaces).isEmpty else {
            return .setup
        }
        if isLoggedIn { return nil }
        return hasSignedInBefore ? .logIn : .welcome
    }

    /// True for the two faces that only need a button.
    public var isOneClick: Bool { self != .setup }
}

File: windows/src/FloatingLyric.Core/LoginPrompt.cs
namespace FloatingLyric.Core;

/// <summary>
/// Which face of the login window is owed to the user. Someone opening the app
/// for the first time should meet a button, not a walkthrough for registering
/// a Spotify app.
/// </summary>
public enum LoginPrompt { Setup, Welcome, LogIn }

public static class LoginPrompts
{
    public static LoginPrompt? Required(string? clientId, bool isLoggedIn, bool hasSignedInBefore)
    {
        if (string.IsNullOrWhiteSpace(clientId)) return LoginPrompt.Setup;
        if (isLoggedIn) return null;
        return hasSignedInBefore ? LoginPrompt.LogIn : LoginPrompt.Welcome;
    }

    public static bool IsOneClick(this LoginPrompt prompt) => prompt != LoginPrompt.Setup;
}

/// <summary>
/// The Spotify app this build ships with, so nobody visits a developer
/// dashboard before hearing their first line. A Client ID is not a secret
/// under PKCE — there is no client secret here at all.
/// </summary>
public static class AppCredentials
{
    public const string BuiltInClientId = "28c043c19e724be68994dc1061251068";

    public static bool HasBuiltInClientId => !string.IsNullOrWhiteSpace(BuiltInClientId);
}
